cjdns-hdr/src/route_header.rs

```diff
@@ -0,0 +1,199 @@
+                Some(from_key)
+            };
+            ip6
+        };
+        // checking invariants
+        if is_ctrl && public_key.is_some() {
+            return Err(HeaderError::CannotParse("public key can not be defined in control frame"));
+        }
+        if is_ctrl && ip6_from_bytes.is_some() {
+            return Err(HeaderError::CannotParse("ip6 is defined for control frame"));
+        }
+        if !is_ctrl && ip6_from_bytes.is_none() {
+            return Err(HeaderError::CannotParse("ip6 is not defined"));
+        }
+        if public_key.is_some() {
+            let ip6_from_key = {
+                let ip6_from_key =
+                    CJDNS_IP6::try_from(public_key.as_ref().expect("zero key bytes")).or(Err(HeaderError::CannotParse("can't create ip6 from public key")))?;
+                Some(ip6_from_key)
+            };
+            if ip6_from_key != ip6_from_bytes {
+                return Err(HeaderError::CannotParse("ip6 derived from public key is not equal to ip6 from header bytes"));
+            }
+        }
+        Ok(RouteHeader {
+            public_key,
+            ip6: ip6_from_bytes,
+            version,
+            switch_header,
+            is_incoming,
+            is_ctrl,
+        })
+    }
+
+    pub fn serialize(&self) -> Result<Vec<u8>> {
+        // checking invariants, because `RouteHeader` can be instantiated without calling constructor
+        if !self.is_ctrl && self.ip6.is_none() {
+            return Err(HeaderError::CannotSerialize("ip6 is is not defined"));
+        }
+        if self.is_ctrl && self.ip6.is_some() {
+            return Err(HeaderError::CannotSerialize("ip6 is defined for control frame"));
+        }
+        if self.is_ctrl && self.public_key.is_some() {
+            return Err(HeaderError::CannotSerialize("public key can not be defined in control frame"));
+        }
+        let public_key_bytes = if self.public_key.is_some() {
+            self.public_key.as_ref().expect("public key is none").bytes()
+        } else {
+            ZERO_PUBLIC_KEY_BYTES.into()
+        };
+        let switch_header_bytes = self.switch_header.serialize()?;
+        let flags = if self.is_ctrl { CONTROL_FRAME } else { INCOMING_FRAME };
+        let pad_bytes = &[0u8; 3];
+        let ip6_bytes = if self.ip6.is_some() {
+            self.ip6.as_ref().expect("ip6 is none").bytes()
+        } else {
+            ZERO_IP6_BYTES.into()
+        };
+
+        let mut data_writer = Writer::with_capacity(ROUTE_HEADER_SIZE);
+        // TODO ask Alex if to follow wrote style in switch and data headers
+        let data = &[
+            public_key_bytes.as_slice(),
+            switch_header_bytes.as_slice(),
+            &self.version.to_be_bytes(),
+            &[flags],
+            pad_bytes,
+            ip6_bytes.as_slice(),
+        ]
+        .concat();
```

Comment:
This is an unusual approach, probably it will require additional allocations. Why not just call `data_writer.write` several times?